### ISSUE TITLE
Cannot extend Map

### ISSUE THREAD

**Issue by philipp-sorin** [mobxjs/mobx#1858]
Hey,
Mobx behaves incorrectly with classes extended from build-ins like Map<K, V>
For example (TS)

```
class StringToStringMap extends Map<string, string> {
  public findOne(id: string): Optional<string> {
    /// code
  }
 }

class Store  {
  @observable
  public stringToStringMap: StringToStringMap  = new StringToStringMap(); 
  // actual type is ObservableMap, method findOne is lost

  @action
  public doStuffWithString(someString: string) {
    this.stringToStringMap.findOne(someString); // throws an error TypeError: findOne is not a function
  } 
}
```

**Comment by fi3ework** [comment]
I test your code and it works fine. My TS version is 3.2.2.

**Comment by philipp-sorin** [comment]
Sounds starag because if test function is written like ```value instanceof Map``` which is satisfy my case

**Comment by mweststrate** [comment]
By default ES6 maps will be converted into observable maps. If you want to
skip that behavior, use `@observable.ref` instead of `@observable`, and the
map itself will not be turned into an observable

Op do 3 jan. 2019 om 13:22 schreef philipp-sorin <notifications@github.com>:

**Comment by philipp-sorin** [comment]
@mweststrate 
But it's not a direct Map, it's a sub type of the Map. 
It should rather throw an Error if it cannot be converted, but silently change subtype to another subtype of base class looks like not good idea. 

Because now, instead of extending Map and creating own observable using Atom, I have to write composite for Map. And this leads to the loss of some features, such as ```instanceof```

**Comment by mweststrate** [comment]
Agreed, PR welcome :) I don't think we had a case before where someone used
a subclass of a Map

Op do 3 jan. 2019 om 14:12 schreef philipp-sorin <notifications@github.com>:

**Comment by philipp-sorin** [comment]
@mweststrate 
Ok, maybe check it later when I will have time, so the issue seems to be closed by now? 
Or should it be a feature request?

**Comment by mweststrate** [comment]
I think we can leave the issue open for now and rename it, to throw in such
cases :)

Op vr 4 jan. 2019 om 11:53 schreef philipp-sorin <notifications@github.com>:

**Comment by mweststrate** [comment]
Addressed and released in 4.9.0 / 5.9.0
